### PR TITLE
CMR-4658: Bulk update for all service revisions

### PR DESCRIPTION
**Pull request by oubiwann**

### CODE REVIEW

File: indexer-app/src/cmr/indexer/services/index_service.clj [hunk 1]
```diff
@@ -101,6 +101,16 @@
                    batch)]
     (es/prepare-batch context (filter-expired-concepts batch) options)))
 
+(defmethod prepare-batch :service
+  [context batch options]
+  ;; Get the service associations as well.
+  (let [batch (map (fn [concept]
+                     (let [service-associations (meta-db/get-associations-for-service
+                                                 context concept)]
+                       (assoc concept :variable-associations service-associations)))
```
Comment (chris-durbin) [inline indexer-app/src/cmr/indexer/services/index_service.clj]: Should be `:service-associations`. Looks like we might be missing a test that verifies associations after bulk index.

File: indexer-app/src/cmr/indexer/services/index_service.clj [hunk 2]
```diff
@@ -422,18 +432,19 @@
...
+  [context concept _parsed-concept options]
   (index-associated-collection context concept options)
   (index-associated-variable context concept {})
   (index-associated-variable context concept {:all-revisions-index? true}))
 
 (defmethod index-concept :service-association
   [context concept parsed-concept options]
-  (index-associated-collection context concept options))
+  (index-associated-collection context concept options)
+  (index-concept context concept parsed-concept {:all-revisions-index? true}))
```
Comment (chris-durbin) [inline indexer-app/src/cmr/indexer/services/index_service.clj]: I'm not sure whether we should be updating associations for the all revisions index. @ygliuvt - do you know what the use case is for associations in the all revisions index?

Also shouldn't this code be `index-associated-service` to the regular index and the all revisions-index? We don't actually index associations, we index the thing they are associated to. Variables does what I expect in lines 440-442 if we do update the all-revisions index for associations (and that would also answer my question about whether we should be indexing to the all revisions index).

This seems to also indicate there's a missing test for creating a service association and verifying that association can be found in the services search route. Do we track associations for services to collections, or are we only tracking associations from collections to services?

File: system-int-test/test/cmr/system_int_test/bootstrap/bulk_index/services_test.clj [hunk 3]
```diff
@@ -93,3 +95,66 @@
          (is (= 3 (count refs)))))
      ;; Re-enable message publishing.
      (core/reenable-automatic-indexing))))
+
+(deftest bulk-index-service-revisions
```
Comment (chris-durbin) [inline system-int-test/test/cmr/system_int_test/bootstrap/bulk_index/services_test.clj]: Note that we are not testing associations at all in this test. Does the variables bulk index test verify associations? From the code it looks like we are tracking associated collections within the variables index, so we should be verifying that in bulk update as well.

File: indexer-app/src/cmr/indexer/services/index_service.clj (hunk 1 above)
Comment (oubiwann) [reply indexer-app/src/cmr/indexer/services/index_service.clj]: Fixed the `:service-associations`. Looking for where we have done any associations tests after bulk index ... have we done this before? Maybe I'm not grepping the right things.

Comment (oubiwann) [reply indexer-app/src/cmr/indexer/services/index_service.clj]: Ah, it seems that there's lots of implicit checking scattered around the sys-int tests for this (maybe that's not good enough? dunno). I had a bunch of failing tests for service associations after I made this change. With my misunderstandings addressed and related updates to my changes, all those failing tests now pass.

File: indexer-app/src/cmr/indexer/services/index_service.clj (hunk 2 above)
Comment (oubiwann) [reply indexer-app/src/cmr/indexer/services/index_service.clj]: I had misunderstood what has happening here; I've just walked through the code again and made corresponding fixes/change. We should continue this conversation based upon those recent changes ...

**Comment by oubiwann** [comment]
Heads-up: @ygliuvt and @chris-durbin I've updated the PR with some changes; could you sanity check this for me? 

@ygliuvt in particular, you have the most accumulated knowledge here ...

File: indexer-app/src/cmr/indexer/services/index_service.clj [hunk 4]
```diff
@@ -401,39 +411,49 @@
...
+  [context concept _parsed-concept options]
   (index-associated-collection context concept options)
   (index-associated-variable context concept {})
   (index-associated-variable context concept {:all-revisions-index? true}))
 
 (defmethod index-concept :service-association
   [context concept parsed-concept options]
-  (index-associated-collection context concept options))
+  (index-associated-collection context concept options)
+  (index-associated-service context concept {})
```
Comment (ygliuvt) [inline indexer-app/src/cmr/indexer/services/index_service.clj]: There is no collection related info in services index, there is no need to do this for services.

Comment (oubiwann) [reply indexer-app/src/cmr/indexer/services/index_service.clj]: Thanks, @ygliuvt!

Turns out this is what was also breaking that one test I couldn't figure out!

**Comment by oubiwann** [comment]
Updates:
 * Tests are all passing locally now
 * Latest changes have been pushed to Github
 * Build is green

File: indexer-app/src/cmr/indexer/services/index_service.clj [hunk 5]
```diff
@@ -101,6 +101,16 @@
                    batch)]
     (es/prepare-batch context (filter-expired-concepts batch) options)))
 
+(defmethod prepare-batch :service
+  [context batch options]
+  ;; Get the service associations as well.
```
Comment (chris-durbin) [inline indexer-app/src/cmr/indexer/services/index_service.clj]: I thought we said we aren't indexing service associations when indexing a service?

File: system-int-test/test/cmr/system_int_test/bootstrap/bulk_index/services_test.clj [hunk 6]
```diff
@@ -93,3 +95,66 @@
...
+           svc1-3 (service/ingest-service svc1-concept)
+           svc2-1 (service/ingest-service svc2-concept)
+           svc2-2 (service/ingest-service svc2-2-concept)
+           svc2-3-tombstone (merge (ingest/delete-concept
+                                    svc2-2-concept (service/token-opts token))
+                                   svc2-2-concept
+                                   {:deleted true
+                                    :user-id "user1"})
+           svc3 (service/ingest-service svc3-concept)]
+       ;; Before bulk indexing, search for variables found nothing
```
Comment (chris-durbin) [inline system-int-test/test/cmr/system_int_test/bootstrap/bulk_index/services_test.clj]: services

File: system-int-test/test/cmr/system_int_test/bootstrap/bulk_index/services_test.clj [hunk 7]
```diff
@@ -93,3 +95,66 @@
...
+       (du/assert-service-umm-jsons-match
+        umm-version/current-service-version
+        []
+        (search/find-concepts-umm-json :service {:all-revisions true}))
+
+       ;; Just index PROV1
+       (bootstrap/bulk-index-services "PROV1")
+       (index/wait-until-indexed)
+
+       ;; After bulk indexing a provider, search found all variable revisions
```
Comment (chris-durbin) [inline system-int-test/test/cmr/system_int_test/bootstrap/bulk_index/services_test.clj]: service

File: system-int-test/test/cmr/system_int_test/bootstrap/bulk_index/services_test.clj [hunk 8]
```diff
@@ -93,3 +95,66 @@
...
+       (du/assert-service-umm-jsons-match
+        umm-version/current-service-version
+        [svc1-1 svc1-2-tombstone svc1-3]
+        (search/find-concepts-umm-json :service {:all-revisions true}))
+
+       ;; Now index all variables
+       (bootstrap/bulk-index-services)
+       (index/wait-until-indexed)
+
+       ;; After bulk indexing, search for variables found all revisions
```
Comment (chris-durbin) [inline system-int-test/test/cmr/system_int_test/bootstrap/bulk_index/services_test.clj]: services

File: system-int-test/test/cmr/system_int_test/bootstrap/bulk_index/services_test.clj [hunk 9]
```diff
@@ -93,3 +95,66 @@
...
+       (index/wait-until-indexed)
+
+       ;; After bulk indexing a provider, search found all variable revisions
+       ;; of that provider
+       (du/assert-service-umm-jsons-match
+        umm-version/current-service-version
+        [svc1-1 svc1-2-tombstone svc1-3]
+        (search/find-concepts-umm-json :service {:all-revisions true}))
+
+       ;; Now index all variables
```
Comment (chris-durbin) [inline system-int-test/test/cmr/system_int_test/bootstrap/bulk_index/services_test.clj]: services

File: indexer-app/src/cmr/indexer/services/index_service.clj [hunk 10]
```diff
@@ -401,32 +411,40 @@
       (let [parsed-coll-concept (cp/parse-concept context coll-concept)]
         (index-concept context coll-concept parsed-coll-concept options)))))
 
-(defn- index-variable
-  "Index the associated variable concept of the given variable concept id."
-  [context variable-concept-id options]
-  (let [var-concept (meta-db/get-latest-concept context variable-concept-id)
-        parsed-var-concept (cp/parse-concept context var-concept)]
-    (index-concept context var-concept parsed-var-concept options)))
+(defn- concept-id->indexed-concept
```
Comment (chris-durbin) [inline indexer-app/src/cmr/indexer/services/index_service.clj]: The arrow syntax usually refers to a data transformation. What about renaming `index-associated-concept`?

Comment (oubiwann) [reply indexer-app/src/cmr/indexer/services/index_service.clj]: Yeah, my thought was very loosely around this: "transforming" a concept id to a proper concept; your suggestion is a good one.

File: indexer-app/src/cmr/indexer/services/index_service.clj (hunk 5 above)
Comment (oubiwann) [reply indexer-app/src/cmr/indexer/services/index_service.clj]: /me nods ...

So it looks like here that would just be a straight-up call to `es/prepare-batch` ... 

/me tries that out.

File: system-int-test/test/cmr/system_int_test/bootstrap/bulk_index/services_test.clj (hunk 6 above)
Comment (oubiwann) [reply system-int-test/test/cmr/system_int_test/bootstrap/bulk_index/services_test.clj]: Fixed.

File: system-int-test/test/cmr/system_int_test/bootstrap/bulk_index/services_test.clj (hunk 7 above)
Comment (oubiwann) [reply system-int-test/test/cmr/system_int_test/bootstrap/bulk_index/services_test.clj]: Fixed.

File: system-int-test/test/cmr/system_int_test/bootstrap/bulk_index/services_test.clj (hunk 9 above)
Comment (oubiwann) [reply system-int-test/test/cmr/system_int_test/bootstrap/bulk_index/services_test.clj]: Fixed.

File: system-int-test/test/cmr/system_int_test/bootstrap/bulk_index/services_test.clj (hunk 8 above)
Comment (oubiwann) [reply system-int-test/test/cmr/system_int_test/bootstrap/bulk_index/services_test.clj]: Fixed.

File: indexer-app/src/cmr/indexer/services/index_service.clj (hunk 10 above)
Comment (oubiwann) [reply indexer-app/src/cmr/indexer/services/index_service.clj]: Changed.

File: indexer-app/src/cmr/indexer/services/index_service.clj (hunk 5 above)
Comment (oubiwann) [reply indexer-app/src/cmr/indexer/services/index_service.clj]: Yup, that works; all changes now pushed.

**Pull request merged**